search/doc_vec.hpp

```diff
@@ -0,0 +1,322 @@
+    std::swap(m_weight, rhs.m_weight);
+  }
+
+  // Returns squared weight of the token-weight pair.
+  double SqrWeight() const { return m_weight * m_weight; }
+
+  Token m_token;
+  double m_weight = 0;
+};
+
+template <typename Token>
+std::string DebugPrint(TokenWeightPair<Token> const & tw)
+{
+  std::ostringstream os;
+  os << "TokenWeightPair [ " << DebugPrint(tw.m_token) << ", " << tw.m_weight << " ]";
+  return os.str();
+}
+
+namespace impl
+{
+// Accumulates weights of equal tokens in |tws|. Result is sorted by
+// tokens. Also, maximum weight from a group of equal tokens will be
+// stored in the corresponding |maxWeight| elem.
+template <typename Token>
+void SortAndMerge(std::vector<TokenWeightPair<Token>> & tws, std::vector<double> & maxWeights)
+{
+  std::sort(tws.begin(), tws.end());
+  size_t n = 0;
+  maxWeights.clear();
+  for (size_t i = 0; i < tws.size(); ++i)
+  {
+    ASSERT_LESS_OR_EQUAL(n, i, ());
+    ASSERT_EQUAL(n, maxWeights.size(), ());
+
+    if (n == 0 || tws[n - 1].m_token != tws[i].m_token)
+    {
+      tws[n].Swap(tws[i]);
+      maxWeights.push_back(tws[n].m_weight);
+      ++n;
+    }
+    else
+    {
+      tws[n - 1].m_weight += tws[i].m_weight;
+      maxWeights[n - 1] = std::max(maxWeights[n - 1], tws[i].m_weight);
+    }
+  }
+
+  ASSERT_LESS_OR_EQUAL(n, tws.size(), ());
+  tws.erase(tws.begin() + n, tws.end());
+}
+
+// Computes squared L2 norm of vector of tokens.
+template <typename Token>
+double SqrL2(std::vector<TokenWeightPair<Token>> const & tws)
+{
+  double sum = 0;
+  for (auto const & tw : tws)
+    sum += tw.SqrWeight();
+  return sum;
+}
+
+// Computes squared L2 norm of vector of tokens + prefix token.
+template <typename Token>
+double SqrL2(std::vector<TokenWeightPair<Token>> const & tws,
+             boost::optional<TokenWeightPair<Token>> const & prefix)
+{
+  double result = SqrL2(tws);
+  return result + (prefix ? prefix->SqrWeight() : 0);
+}
+}  // namespace impl
+
+// This class represents a document in a vector space of tokens.
+template <typename Token>
+class DocVec
+{
+public:
+  using TokenWeightPair = TokenWeightPair<Token>;
+  using TokenWeightPairs = std::vector<TokenWeightPair>;
+
+  class Builder
+  {
+  public:
+    template <typename T>
+    void Add(T && token, double weight)
+    {
+      m_tws.emplace_back(std::forward<T>(token), weight);
+    }
+
+  private:
+    friend class DocVec;
+
+    TokenWeightPairs m_tws;
+  };
+
+  DocVec() = default;
+  explicit DocVec(Builder && builder) : m_tws(std::move(builder.m_tws)) { Init(); }
+  explicit DocVec(Builder const & builder) : m_tws(builder.m_tws) { Init(); }
+
+  TokenWeightPairs const & GetTokenWeightPairs() const { return m_tws; }
+  std::vector<double> const & GetMaxWeights() const { return m_maxWeights; }
+
+  bool Empty() const { return m_tws.empty(); }
+
+private:
+  template <typename T>
+  friend std::string DebugPrint(DocVec<T> const & dv)
+  {
+    return "DocVec " + DebugPrint(dv.m_tws);
```

Comment:
[] ?

Reply:
No, it's okay now - m_tws is a vector, so it will be already wrapped into square brackets.